add watches on injection points

natasha41575 opened the pull request:
Had to dig around a bit to figure out how to add dynamic watches, please feel free to let me know if there is a more conventional way to do it :)

Tested it with https://github.com/nephio-test/natasha41575-edge-1/blob/drafts/default/packagedeployment-1/default/cluster-context.yaml - I can see the spec in this file gets updated whenever I update the cluster-context resource in my local cluster.

Fixes https://github.com/nephio-project/one-summit-22-workshop/issues/38

natasha41575 committed:
add watches on injection points

johnbelamaric commented:
Oo, very nice. So, now if we for example change the cluster scale profile, we will get a new (or update an existing) Draft with the new value injected?

That's super cool. One thing I think we should add before we go wild with updating injected objects is a way to verify that there has been a change. One complication is that a function may have changed the in-package spec of the injected resource. So, here is what I propose (in this PR):

- Recall that we only inject the "spec". This means the metadata of the in-package resource remains usable. So, we will take advantage of that.
- Before injecting the spec object, we calculate a hash for it. If it doesn't already exist, we add an annotation `automation.nephio.org/injection-hash` with that hash. Otherwise we compare the hashes.
- If this is a first injection or the hashes are different, we go ahead with the spec replacement. Otherwise we do nothing.

Can you also take time to make this into an "injector" library of sorts. You can see we are doing the same thing here: https://github.com/henderiw-nephio/nf-injector-controller/blob/main/controllers/injector/reconciler.go

natasha41575 committed:
use hash and refactor

natasha41575 commented:
> Oo, very nice. So, now if we for example change the cluster scale profile, we will get a new (or update an existing) Draft with the new value injected?

Originally, it would just update the existing draft. But I updated it so that if there is no draft, and the existing revision is a published one, it will create a new draft revision with the new value injected.

> Before injecting the spec object, we calculate a hash for it. If it doesn't already exist, we add an annotation automation.nephio.org/injection-hash with that hash. Otherwise we compare the hashes. If this is a first injection or the hashes are different, we go ahead with the spec replacement. Otherwise we do nothing.

Done.

> Can you also take time to make this into an "injector" library of sorts.

Happy to do so! But I would prefer to do it in a followup PR if that's alright - I'd like to take a little more time to think about the best way to structure such a library. I refactored parts of the code in this PR to make it more readable.

johnbelamaric commented:
Yes, library can be a different PR or maybe even a different repo!

I will look at this when I am back after the Thanksgiving holiday.

johnbelamaric commented on controllers/automation/packagedeployment_controller.go:
```diff
@@ -271,31 +304,74 @@ metadata:
+					APIVersion: clusterResource.GetAPIVersion(),
+					Kind:       clusterResource.GetKind(),
+				},
+				NameMeta: yaml.NameMeta{
```
As I understand this, it is watching the specific resources, by name. If we have 1000s of clusters and 100s of packages, that could result in a LOT of watches. I think it would be more efficient to watch just on the TypeMeta, then keep an internal mapping. I will create a separate issue for that though. As is works for our current prototyping.
